Source/WebInspectorUI/UserInterface/Controllers/DeviceSettingsManager.js

```diff
@@ -0,0 +1,178 @@
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+WI.DeviceSettingsManager = class DeviceSettingsManager extends WI.Object
+{
+    constructor()
+    {
+        super();
+
+        this._deviceUserAgent = "";
+        this._overridenDeviceUserAgent = null;
+        this._overridenDeviceScreenSize = null;
+        this._overridenDeviceSettings = new Map;
+    }
+
+    // Target
+
+    initializeTarget(target)
+    {
+        if (!target.hasDomain("Page"))
+            return
+
+        // COMPATIBILITY (iOS 12.2): Page.overrideUserAgent did not exist.
+        if (target.hasCommand("Page.overrideUserAgent") && this._overridenDeviceUserAgent)
+            target.PageAgent.overrideUserAgent(this._overridenDeviceUserAgent);
+
+        // COMPATIBILITY (iOS 12.2): Page.overrideSetting did not exist.
+        if (target.hasCommand("Page.overrideSetting")) {
+            for (let [setting, value] of this._overridenDeviceSettings)
+                target.PageAgent.overrideSetting(setting, value);
+        }
+
+        WI.runtimeManager.evaluateInInspectedWindow("globalThis.navigator?.userAgent", {objectGroup: WI.DeviceSettingsManager.ObjectGroup, returnByValue: true}, (remoteObject, wasThrown, result) => {
+            if (!wasThrown && result)
+                this._deviceUserAgent = result.value;
+
+            target.RuntimeAgent.releaseObjectGroup(WI.DeviceSettingsManager.ObjectGroup);
+        });
+    }
+
+    // Static
+
+    static supportsSetScreenSizeOverride()
+    {
+        return InspectorBackend.hasCommand("Page.setScreenSizeOverride");
+    }
+
+    // Public
+
+    get deviceUserAgent() { return this._deviceUserAgent; }
+    get overridenDeviceUserAgent() { return this._overridenDeviceUserAgent; }
+    get overridenDeviceScreenSize() { return this._overridenDeviceScreenSize; }
+    get overridenDeviceSettings() { return this._overridenDeviceSettings; }
+
+    get hasOverridenDefaultSettings() {
+        return this._overridenDeviceUserAgent || this._overridenDeviceScreenSize || this._overridenDeviceSettings.size > 0;
+    }
+
+    overrideDeviceSetting(setting, value, callback)
+    {
+        let target = WI.assumingMainTarget();
+        if (!target.hasCommand("Page.overrideSetting")) {
+            callback(false);
+            return;
+        }
+
+        let commandArguments = {
+            setting,
+        };
+
+        let shouldOverride = !this._overridenDeviceSettings.has(setting)
+            commandArguments.value = value; // Providing `value` applies the override for `setting`. Not providing it clears the override.
```

Comment:
i think you're missing a `if (shouldOverride)`

NIT: this comment doesn't really add anything as it's a pretty common pattern within Web Inspector where not providing one of the arguments acts as a clear (e.g.

Reply:
>i think you're missing a if (shouldOverride)

Good catch! 

>NIT: this comment doesn't really add anything

Fine, I'll remove the comment.